File: examples/algorithms/004-binary-search/implementations/python/binary_search.py
#!/usr/bin/env python3
"""
Binary Search - Python Implementation
Ergonomic baseline for Ruchy comparison
"""

from typing import List, Optional, TypeVar, Generic
import time
from collections import defaultdict

T = TypeVar('T')

class BinarySearcher(Generic[T]):
    """Enhanced binary search with caching and performance tracking"""
    
    def __init__(self, data: List[T]):
        self.data = sorted(data)  # Ensure data is sorted
        self.cache = {}
        self.stats = {
            'cache_hits': 0,
            'cache_misses': 0,
            'total_comparisons': 0,
            'total_searches': 0
        }
    
    def binary_search(self, target: T) -> Optional[int]:
        """Basic binary search implementation"""
        left, right = 0, len(self.data)
        comparisons = 0
        
        while left < right:
            comparisons += 1
            mid = left + (right - left) // 2
            
            if self.data[mid] == target:
                self.stats['total_comparisons'] += comparisons
                return mid
            elif self.data[mid] < target:
                left = mid + 1
            else:
                right = mid
        
        self.stats['total_comparisons'] += comparisons
        return None
    
    def search_with_cache(self, target: T) -> Optional[int]:
        """Enhanced search with caching"""
        self.stats['total_searches'] += 1
        
        # Check cache first
        if target in self.cache:
            self.stats['cache_hits'] += 1
            print(f"🎯 Cache hit for target: {target}")
            return self.cache[target]
        
        # Cache miss - perform search
        self.stats['cache_misses'] += 1
        result = self.binary_search(target)
        
        # Cache the result
        self.cache[target] = result
        print(f"💾 Cached result for {target}: {result}")
        
        return result
    
    def binary_search_leftmost(self, target: T) -> Optional[int]:
        """Find leftmost occurrence of target"""
        left, right = 0, len(self.data)
        
        while left < right:
            mid = left + (right - left) // 2
            
            if self.data[mid] < target:
                left = mid + 1
            else:
                right = mid
        
        if left < len(self.data) and self.data[left] == target:
            return left
        return None
    
    def binary_search_rightmost(self, target: T) -> Optional[int]:
        """Find rightmost occurrence of target"""
        left, right = 0, len(self.data)
        
        while left < right:
            mid = left + (right - left) // 2
            
            if self.data[mid] <= target:
                left = mid + 1
            else:
                right = mid
        
        if left > 0 and self.data[left - 1] == target:
            return left - 1
        return None
    
    def search_range(self, start: T, end: T) -> List[int]:
        """Find all indices in range [start, end]"""
        left = self.binary_search_leftmost(start)
        right = self.binary_search_rightmost(end)
        
        if left is None or right is None:
            return []
        
        return list(range(left, right + 1))
    
    def show_performance(self):
        """Display performance statistics"""
        total = self.stats['total_searches']
        hits = self.stats['cache_hits']
        misses = self.stats['cache_misses']
        comparisons = self.stats['total_comparisons']
        
        hit_rate = (hits / total * 100) if total > 0 else 0
        avg_comparisons = (comparisons / misses) if misses > 0 else 0
        
        print(f"\n📊 Performance Statistics:")
        print(f"   Total searches: {total}")
        print(f"   Cache hits: {hits} ({hit_rate:.1f}%)")
        print(f"   Cache misses: {misses}")
        print(f"   Total comparisons: {comparisons}")
        if misses > 0:
            print(f"   Avg comparisons per search: {avg_comparisons:.1f}")

def binary_search_simple(arr: List[T], target: T) -> Optional[int]:
    """Standalone binary search function"""
    left, right = 0, len(arr)
    
    while left < right:
        mid = left + (right - left) // 2
        
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid
    
    return None

def benchmark_search(data: List[int], targets: List[int], iterations: int = 10000) -> float:
    """Benchmark binary search performance"""
    start_time = time.perf_counter()
    
    for _ in range(iterations):
        for target in targets:
            binary_search_simple(data, target)
    
    end_time = time.perf_counter()
    total_operations = iterations * len(targets)
    
    return (end_time - start_time) * 1_000_000_000 / total_operations  # ns per operation

def interpolation_search(arr: List[int], target: int) -> Optional[int]:
    """Interpolation search for uniformly distributed data"""
    left, right = 0, len(arr) - 1
    
    while left <= right and target >= arr[left] and target <= arr[right]:
        if left == right:
            return left if arr[left] == target else None
        
        # Interpolation formula
        pos = left + int(((target - arr[left]) * (right - left)) / (arr[right] - arr[left]))
        pos = max(left, min(pos, right))  # Ensure pos is within bounds
        
        if arr[pos] == target:
            return pos
        elif arr[pos] < target:
            left = pos + 1
        else:
            right = pos - 1
    
    return None

def main():
    print("🔍 Binary Search - Python Implementation")
    print("========================================")
    
    # Test data
    data = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 23, 29, 31, 37, 41, 43, 47]
    print(f"📊 Test data ({len(data)} elements): {data}")
    print()
    
    # Test basic binary search
    print("🔍 Basic binary search tests:")
    for target in [5, 11, 20, 1, 47, 25]:
        result = binary_search_simple(data, target)
        if result is not None:
            print(f"   ✅ Found {target} at index {result}")
        else:
            print(f"   ❌ {target} not found")
    
    # Test boundary searches with duplicates
    print("\n🔄 Boundary search tests (with duplicates):")
    duplicates = [1, 2, 2, 2, 3, 5, 5, 7, 7, 7, 9]
    searcher_dup = BinarySearcher(duplicates)
    print(f"   Data with duplicates: {duplicates}")
    
    for target in [2, 5, 7]:
        leftmost = searcher_dup.binary_search_leftmost(target)
        rightmost = searcher_dup.binary_search_rightmost(target)
        print(f"   {target} -> leftmost: {leftmost}, rightmost: {rightmost}")
    
    # Test range queries
    print(f"\n📏 Range query tests:")
    range_indices = searcher_dup.search_range(2, 7)
    print(f"   Elements in range [2, 7]: indices {range_indices}")
    range_values = [duplicates[i] for i in range_indices]
    print(f"   Values: {range_values}")
    
    print("\n🚀 Enhanced search with caching:")
    
    # Test enhanced searcher with caching
    searcher = BinarySearcher(data)
    
    # Demonstrate caching behavior
    test_targets = [7, 15, 7, 23, 15, 7, 41, 23]
    for target in test_targets:
        result = searcher.search_with_cache(target)
        if result is not None:
            print(f"   ✅ Found {target} at index {result}")
        else:
            print(f"   ❌ {target} not found")
    
    # Show performance statistics
    searcher.show_performance()
    
    # Test interpolation search
    print(f"\n🔬 Interpolation search test:")
    uniform_data = list(range(0, 100, 2))  # [0, 2, 4, 6, ..., 98]
    print(f"   Uniform data sample: {uniform_data[:10]} ... {uniform_data[-10:]}")
    
    for target in [20, 50, 77, 99]:
        result = interpolation_search(uniform_data, target)
        if result is not None:
            print(f"   ✅ Interpolation found {target} at index {result}")
        else:
            print(f"   ❌ Interpolation: {target} not found")
    
    # Benchmark performance
    print(f"\n⚡ Performance benchmark:")
    benchmark_targets = [5, 11, 15, 23, 29, 37, 99]  # Mix of found/not found
    avg_time = benchmark_search(data, benchmark_targets, 10000)
    print(f"   Average search time: {avg_time:.2f} ns per operation")
    
    print(f"\n✅ Python binary search demonstration complete")
    print(f"🐍 Ergonomic baseline established for Ruchy comparison")

if __name__ == "__main__":
    main()

File: examples/algorithms/004-binary-search/implementations/rust/binary_search.rs
// Binary Search - Rust Implementation
// Baseline performance reference for Ruchy comparison

use std::collections::HashMap;
use std::cmp::Ordering;
use std::time::Instant;

// Basic binary search function
fn binary_search<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len();
    
    while left < right {
        let mid = left + (right - left) / 2;
        
        match arr[mid].cmp(target) {
            Ordering::Equal => return Some(mid),
            Ordering::Less => left = mid + 1,
            Ordering::Greater => right = mid,
        }
    }
    
    None
}

// Enhanced binary search with caching for performance comparison
struct BinarySearcher<T: Ord + Clone + std::hash::Hash> {
    cache: HashMap<T, Option<usize>>,
    data: Vec<T>,
    stats: SearchStats,
}

#[derive(Debug, Default)]
struct SearchStats {
    cache_hits: usize,
    cache_misses: usize,
    total_comparisons: usize,
}

impl<T: Ord + Clone + std::hash::Hash + std::fmt::Debug> BinarySearcher<T> {
    fn new(data: Vec<T>) -> Self {
        Self {
            cache: HashMap::new(),
            data,
            stats: SearchStats::default(),
        }
    }
    
    fn search_with_cache(&mut self, target: &T) -> Option<usize> {
        // Check cache first
        if let Some(&cached_result) = self.cache.get(target) {
            self.stats.cache_hits += 1;
            println!("🎯 Cache hit for target: {:?}", target);
            return cached_result;
        }
        
        // Cache miss - perform search with comparison counting
        self.stats.cache_misses += 1;
        let result = self.binary_search_with_stats(target);
        
        // Cache the result
        self.cache.insert(target.clone(), result);
        println!("💾 Cached result for {:?}: {:?}", target, result);
        
        result
    }
    
    fn binary_search_with_stats(&mut self, target: &T) -> Option<usize> {
        let mut left = 0;
        let mut right = self.data.len();
        
        while left < right {
            self.stats.total_comparisons += 1;
            let mid = left + (right - left) / 2;
            
            match self.data[mid].cmp(target) {
                Ordering::Equal => return Some(mid),
                Ordering::Less => left = mid + 1,
                Ordering::Greater => right = mid,
            }
        }
        
        None
    }
    
    fn show_performance(&self) {
        let total_searches = self.stats.cache_hits + self.stats.cache_misses;
        let cache_hit_rate = if total_searches > 0 {
            self.stats.cache_hits as f64 / total_searches as f64 * 100.0
        } else {
            0.0
        };
        
        println!("\n📊 Performance Statistics:");
        println!("   Total searches: {}", total_searches);
        println!("   Cache hits: {} ({:.1}%)", self.stats.cache_hits, cache_hit_rate);
        println!("   Cache misses: {}", self.stats.cache_misses);
        println!("   Total comparisons: {}", self.stats.total_comparisons);
        if self.stats.cache_misses > 0 {
            println!("   Avg comparisons per search: {:.1}", 
                     self.stats.total_comparisons as f64 / self.stats.cache_misses as f64);
        }
    }
}

// Specialized search variants for comprehensive comparison
fn binary_search_leftmost<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len();
    
    while left < right {
        let mid = left + (right - left) / 2;
        
        if arr[mid] < *target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    if left < arr.len() && arr[left] == *target {
        Some(left)
    } else {
        None
    }
}

fn binary_search_rightmost<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len();
    
    while left < right {
        let mid = left + (right - left) / 2;
        
        if arr[mid] <= *target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    if left > 0 && arr[left - 1] == *target {
        Some(left - 1)
    } else {
        None
    }
}

fn benchmark_search(data: &[i32], targets: &[i32], iterations: usize) -> f64 {
    let start = Instant::now();
    
    for _ in 0..iterations {
        for &target in targets {
            binary_search(data, &target);
        }
    }
    
    start.elapsed().as_nanos() as f64 / (iterations * targets.len()) as f64
}

fn main() {
    println!("🔍 Binary Search - Rust Baseline Implementation");
    println!("===============================================");
    
    let data = vec![1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 23, 29, 31, 37, 41, 43, 47];
    println!("📊 Test data ({} elements): {:?}", data.len(), data);
    println!();
    
    // Test basic binary search
    println!("🔍 Basic binary search tests:");
    for target in &[5, 11, 20, 1, 47, 25] {
        match binary_search(&data, target) {
            Some(idx) => println!("   ✅ Found {} at index {}", target, idx),
            None => println!("   ❌ {} not found", target),
        }
    }
    
    // Test specialized searches for duplicates
    println!("\n🔄 Boundary search tests (with duplicates):");
    let duplicates = vec![1, 2, 2, 2, 3, 5, 5, 7, 7, 7, 9];
    println!("   Data with duplicates: {:?}", duplicates);
    
    for target in &[2, 5, 7] {
        let leftmost = binary_search_leftmost(&duplicates, target);
        let rightmost = binary_search_rightmost(&duplicates, target);
        println!("   {} -> leftmost: {:?}, rightmost: {:?}", target, leftmost, rightmost);
    }
    
    println!("\n🚀 Enhanced search with caching:");
    
    // Test enhanced searcher with caching
    let mut searcher = BinarySearcher::new(data.clone());
    
    // Demonstrate caching behavior
    let test_targets = vec![7, 15, 7, 23, 15, 7, 41, 23];
    for target in test_targets {
        match searcher.search_with_cache(&target) {
            Some(idx) => println!("   ✅ Found {} at index {}", target, idx),
            None => println!("   ❌ {} not found", target),
        }
    }
    
    // Show performance statistics
    searcher.show_performance();
    
    // Benchmark performance
    println!("\n⚡ Performance benchmark:");
    let benchmark_targets = vec![5, 11, 15, 23, 29, 37, 99]; // Mix of found/not found
    let avg_time = benchmark_search(&data, &benchmark_targets, 10000);
    println!("   Average search time: {:.2} ns per operation", avg_time);
    
    println!("\n✅ Rust binary search demonstration complete");
    println!("🎯 Baseline established for Ruchy performance comparison");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_binary_search_basic() {
        let arr = [1, 3, 5, 7, 9];
        
        assert_eq!(binary_search(&arr, &5), Some(2));
        assert_eq!(binary_search(&arr, &1), Some(0));
        assert_eq!(binary_search(&arr, &9), Some(4));
        assert_eq!(binary_search(&arr, &4), None);
        assert_eq!(binary_search(&arr, &10), None);
        
        // Test empty array
        let empty: [i32; 0] = [];
        assert_eq!(binary_search(&empty, &5), None);
    }
    
    #[test]
    fn test_boundary_searches() {
        let arr = [1, 2, 2, 2, 3, 5, 5];
        
        assert_eq!(binary_search_leftmost(&arr, &2), Some(1));
        assert_eq!(binary_search_rightmost(&arr, &2), Some(3));
        assert_eq!(binary_search_leftmost(&arr, &5), Some(5));
        assert_eq!(binary_search_rightmost(&arr, &5), Some(6));
        assert_eq!(binary_search_leftmost(&arr, &4), None);
    }
    
    #[test]
    fn test_cached_search() {
        let data = vec![1, 3, 5, 7, 9];
        let mut searcher = BinarySearcher::new(data);
        
        // First search should be cache miss
        assert_eq!(searcher.search_with_cache(&5), Some(2));
        assert_eq!(searcher.stats.cache_misses, 1);
        assert_eq!(searcher.stats.cache_hits, 0);
        
        // Second search should be cache hit
        assert_eq!(searcher.search_with_cache(&5), Some(2));
        assert_eq!(searcher.stats.cache_misses, 1);
        assert_eq!(searcher.stats.cache_hits, 1);
    }
}